internal/gardenclient/shoot_client.go

```diff
@@ -0,0 +1,273 @@
+}
+
+// generate generates a Kubernetes kubeconfig for communicating with the kube-apiserver
+// by exec'ing the gardenlogin plugin, which fetches a client certificate.
+// If legacy is false, the shoot reference and garden cluster identity is passed via the cluster extensions,
+// which is supported starting with kubectl version v1.20.0.
+// If legacy is true, the shoot reference and garden cluster identity are passed as command line flags to the plugin
+func (k *shootKubeconfigRequest) generate(legacy bool) ([]byte, error) {
+	authName := fmt.Sprintf("%s--%s", k.namespace, k.shootName) // TODO instead of namespace, use project? But this would require an additional call
+	name := fmt.Sprintf("%s-%s", authName, k.clusters[0].name)
+
+	var legacyArgs []string
+	if legacy {
+		legacyArgs = []string{
+			fmt.Sprintf("--name=%s", k.shootName),
+			fmt.Sprintf("--namespace=%s", k.namespace),
+			fmt.Sprintf("--garden-cluster-identity=%s", k.gardenClusterIdentity),
+		}
+	}
+
+	var authInfos []clientcmdv1.NamedAuthInfo
+	authInfos = append(authInfos, clientcmdv1.NamedAuthInfo{
+		Name: authName,
+		AuthInfo: clientcmdv1.AuthInfo{
+			Exec: &clientcmdv1.ExecConfig{
+				Command: "kubectl",
+				Args: append([]string{
+					"gardenlogin",
+					"get-client-certificate",
+				},
+					legacyArgs...,
+				),
+				Env:                nil,
+				APIVersion:         clientauthenticationv1beta1.SchemeGroupVersion.String(),
+				InstallHint:        "",
+				ProvideClusterInfo: true,
+			},
+		},
+	})
+
+	config := &clientcmdv1.Config{
+		CurrentContext: name,
+		Clusters:       []clientcmdv1.NamedCluster{},
+		Contexts:       []clientcmdv1.NamedContext{},
+		AuthInfos:      authInfos,
+	}
+
+	extension := ExecPluginConfig{
+		ShootRef: ShootRef{
+			Namespace: k.namespace,
+			Name:      k.shootName,
+		},
+		GardenClusterIdentity: k.gardenClusterIdentity,
+	}
+
+	var clusterExtensions []clientcmdv1.NamedExtension
+
+	if !legacy {
+		raw, err := json.Marshal(extension)
+		if err != nil {
+			return nil, fmt.Errorf("could not json marshal cluster extension: %w", err)
+		}
+
+		clusterExtensions = []clientcmdv1.NamedExtension{
+			{
+				Name: "client.authentication.k8s.io/exec",
+				Extension: runtime.RawExtension{
+					Raw: raw,
+				},
+			},
+		}
+	}
+
+	for _, cluster := range k.clusters {
+		name := fmt.Sprintf("%s-%s", authName, cluster.name)
+
+		config.Clusters = append(config.Clusters, clientcmdv1.NamedCluster{
+			Name: name,
+			Cluster: clientcmdv1.Cluster{
+				CertificateAuthorityData: cluster.caCert,
+				Server:                   fmt.Sprintf("https://%s", cluster.apiServerHost),
+				Extensions:               clusterExtensions,
+			},
+		})
+		config.Contexts = append(config.Contexts, clientcmdv1.NamedContext{
+			Name: name,
+			Context: clientcmdv1.Context{
+				Cluster:   name,
+				AuthInfo:  authName,
+				Namespace: "default", // TODO leave hardcoded? Or omit?
+			},
+		})
+	}
+
+	return runtime.Encode(clientcmdlatest.Codec, config)
+}
```

Comment:
```suggestion
func (k *shootKubeconfigRequest) generate(legacy bool) ([]byte, error) {
	var extension *ExecPluginConfig

	args := []string{
		"gardenlogin",
		"get-client-certificate",
	}

	if legacy {
		args = append(
			args,
			fmt.Sprintf("--name=%s", k.shootName),
			fmt.Sprintf("--namespace=%s", k.namespace),
			fmt.Sprintf("--garden-cluster-identity=%s", k.gardenClusterIdentity),
		)
	} else {
		extension = &ExecPluginConfig{
			ShootRef: ShootRef{
				Namespace: k.namespace,
				Name:      k.shootName,
			},
			GardenClusterIdentity: k.gardenClusterIdentity,
		}
	}

	config := clientcmdapi.NewConfig()

	authInfo := clientcmdapi.NewAuthInfo()
	authInfo.Exec = &clientcmdapi.ExecConfig{
		Command:            "kubectl",
		Args:               args,
		Env:                nil,
		APIVersion:         clientauthenticationv1beta1.SchemeGroupVersion.String(),
		InstallHint:        "",
		ProvideClusterInfo: true,
	}
	authName := fmt.Sprintf("%s--%s", k.namespace, k.shootName) // TODO instead of namespace, use project? But this would require an additional call
	config.AuthInfos[authName] = authInfo

	for i, c := range k.clusters {
		name := fmt.Sprintf("%s-%s", authName, c.name)
		if i == 0 {
			config.CurrentContext = name
		}

		cluster := clientcmdapi.NewCluster()
		cluster.CertificateAuthorityData = c.caCert
		cluster.Server = fmt.Sprintf("https://%s", c.apiServerHost)
		if !legacy {
			cluster.Extensions["client.authentication.k8s.io/exec"] = extension
		}
		config.Clusters[name] = cluster

		context := clientcmdapi.NewContext()
		context.Cluster = name
		context.AuthInfo = authName
		context.Namespace = "default" // TODO leave hardcoded? Or omit?
		config.Contexts[name] = context
	}

	return clientcmd.Write(*config)
}
```
I am not the expert here but I would use the `clientcmdapi` to create the kubeconfig.